apptainer/apptainer.qmd

```diff
@@ -2,19 +2,121 @@
+- For practical purposes, they are the same
 
-## Portability to HPC systems, & reproducibilty note: .sif files vs rebuilding from a .def
+## Structure of an apptainer command
+
+- `apptainer [subcommand] [image] [additional commands]` 
+
+- Example of `pull` and `shell`
+
+```
+apptainer pull docker://alpine
+apptainer shell alpine_latest.sif
+cat /etc/alpine-release
+exit
+cat /etc/alpine-release
+cat /etc/debian_version
+```
+
+- Interact with the container from the host system:
+
+```
+apptainer exec alpine_latest.sif cat /etc/alpine-release
+cat /etc/alpine-release
+```
+## Building .sif from a definition file (.def)
+
+- Building a custom container requires a .def file, specifying the registry and image for the base image, and [various options for the container](https://apptainer.org/documentation/)
+
+```default{filename="example.def"}
+
+Bootstrap: docker
+From: debian:12.5-slim
+
+%environment
+	export PATH=$PATH:/root/.pixi/bin
+
+%runscript
+	cat /etc/debian_version
+
+%post
+	export PATH=$PATH:/root/.pixi/bin
+	apt-get update && \
+	apt-get install -y curl && \
+	curl -fsSL https://pixi.sh/install.sh | bash && \
+	apt-get clean && \
+	pixi global install -c bioconda -c conda-forge  minigraph
+
+```
+
+- Then build the container: `apptainer build minigraph.sif example.def`
+
+- `apptainer run` executes the runscript inside the container:
+
+`apptainer run minigraph.sif`
+
+## Portability to HPC systems, & note on reproducibilty (.sif files vs rebuilding from a .def)
+
+- .sif files are portable and will be highly reproducible
+
+- You can also share a .def file, however consider that some tags on docker hub refer to rolling releases, e.g. `latest`. If you want future builds of the container to be identical, try to find a static tag, e.g. a github commit tag
+
+## Other useful things to know:
+
+### Mount binding
+
+- Some folders are automatically bound from the host system (e.g., `$HOME`, `$CWD`, `/tmp`)
+- Therefore don't install software to those locations - they'll install on your host system too
+- Use an unmounted folder like `/opt` or `/usr/local` instead
+- If you need to mount a directory to the container, e.g. a data directory, this is possible, e.g.:
+
+`apptainer exec --bind /scratch example.sif ls /scratch`
+
+### Conversion from docker
+- It's possible to convert from docker images to singularity images - not covered today
+
+### Sandbox containers
+
+- Singularity containers are basically uneditable - no so fun if you need to keep rebuilding from scatch during development
+- There is a "sandbox container" feature which is editable, and it works like a file system within your file system
+- When you are ready for a production container, you can convert the sandbox container to a regular container, though it would be preferable to rebuild from a definition file, so that there is a record of what was done
```

Comment:
Perhaps have some callout boxes with tips. 

The way I work with containers is to never have a big container for everything. Rebuilding containers is time-wasting, especially if you need something that conflicts with a previous tool. The container has at most what it needs for a script/Nextflow process. I'll almost always look at nf-core modules to see if something there has what I need already. 

When prototyping, if a container is not yet available, for example when you need to load various r-packages, I would do that in a conda/mamba/pixi environment first, and then make a container at the end ( now with seqera containers / although perhaps also mention this is based on biocontainers initiative). It's not fool proof, because it's quite likely one might go back and edit more, but it saves building a container over and over again and debugging that.